lib/rouge/lexers/solidity.rb

```diff
@@ -0,0 +1,182 @@
+# -*- coding: utf-8 -*- #
+
+module Rouge
+  module Lexers
+    class Solidity < RegexLexer
+      tag 'solidity'
+      filenames '*.sol', '*.solidity'
+      mimetypes 'text/x-solidity'
+
+      title "Solidity"
+      desc "Solidity, an Ethereum smart contract programming language"
+
+      # optional comment or whitespace
+      ws = %r((?:\s|//.*?\n|/[*].*?[*]/)+)
+      id = /[a-zA-Z_][a-zA-Z0-9_]*/
+
+
+      def self.analyze_text(text)
+        return 1 if text.shebang? 'pragma solidity'
+      end
+
+      # TODO: seperate by "type"
+      def self.keywords
+        @keywords ||= Set.new %w(
+          anonymous as assembly break constant continue contract do delete
+          else enum event external for function hex if indexed interface
+          internal import is library mapping memory modifier new payable
+          public pragma private return returns storage struct throw
+          using var while
+        )
+      end
+
+      def self.builtins
+        @builtins ||= Set.new %w(
+          now
+          true false
+          assert require revert
+          selfdestruct suicide
+          this super balance transfer send call callcode delegatecall
+          addmod mulmod keccak256 sha256 sha3 ripemd160 ecrecover
+        )
+        # TODO: use (currently shadowed by catch-all in :statements)
+        block = %w(blockhash coinbase difficulty gaslimit number timestamp)
+        @builtins.merge( block.map { |i| "block.#{i}" } )
+        msg = %w(data gas sender sig value)
+        @builtins.merge( msg.map { |i| "msg.#{i}" } )
+        tx = %w(gasprice origin)
+        @builtins.merge( tx.map { |i| "tx.#{i}" } )
+      end
+
+      def self.constants
+        @constants ||= Set.new %w(
+          wei finney szabo ether
+          seconds minutes hours days weeks years
+        )
+      end
+
+      def self.keywords_type
+        @keywords_type ||= Set.new %w(
+          int uint bytes fixed ufixed address bool
+        )
+
+        # bytes1 .. bytes32
+        @keywords_type.merge( (1..32).map { |i| "bytes#{i}" } )
```

Comment:
Unfortunately this and all the following will be executed every time `self.keywords_type` is called - i.e. nearly every time an identifier is scanned. I suggest moving the code into a `@keywords_type ||= begin ... end` style block, so that the `||=` protects the entirety of it from being executed more than once. Alternatively, adding `return @keywords_type if @keywords_type` at the top and getting rid of the `||` above.